hardware/measurement/adc_stream_rate.py

```
@@ -0,0 +1,448 @@
+    delivered = sum(b.nsamples for b in used)
+    taken_delta = (last.taken - first.taken) % (1 << 32)
+    dropped_delta = (last.dropped - first.dropped) % (1 << 16)
+
+    # taken / dropped は「差」なので、比べる相手も同じ区間に揃える。
+    # first block 自身の sample は first.taken の時点で既に数え終わっているため、
+    # 区間内に届いたのは先頭 block を除いたぶんである。ここを揃えないと
+    # 先頭 block ぶんだけずれた比較になる。
+    delivered_in_interval = delivered - first.nsamples
+
+    # 区間内の収支。ISRが取得した数は、届いた数 + ISRが捨てた数 + 回線上で失った数。
+    # 差が残るなら block が回線上で失われている（block欠番と符合するはず）。
```

Ruff (0.16.1)

[warning] Comment contains ambiguous `（` (FULLWIDTH LEFT PARENTHESIS). Did you mean `(` (LEFT PARENTHESIS)?

(RUF003)